gen/app-gen

```diff
@@ -0,0 +1,307 @@
+        docker pull "$url" --platform "$platform"
+    else
+        ctr image pull --platform "$platform"
+    fi
+}
+
+export_image() {
+    local -r image="$1"
+    local -r output="$2"
+
+    [[ "$image" == "" ]] && { return 0; }
+    if [[ "${orchestrator}" == "docker-compose" ]]; then
+        docker image save "$image" -o "${output}"
+    else
+        ctr image export "${output}" "$image" --platform "$platform"
+    fi
+}
+
+prepare_images() {
+    local -r root_dir="$1"
+    local i
+    local j
+    local url_current
+    local url_new
+    local output_dir
+
+    for ((i = 0; i < ${#images[@]}; i++)); do
+        pull_image "${images[${i}]}"
+        images_sha+=($(get_image_sha "${images[${i}]}"))
+    done
+    declare -p images_sha
+
+    for ((i = 0; i < ${#images[@]}; i += 2)); do
+        j=$((i + 1))
+        url_current=${images[${i}]}
+        url_new=${images[${j}]}
+        output_dir="${root_dir}/${images_sha[${j}]}"
+        [[ -d "$output_dir" ]] || mkdir "$output_dir"
+        echo "${url_new}" > "${output_dir}/url-new.txt"
+        echo "${url_current}" > "${output_dir}/url-current.txt"
+        echo "${images_sha[${i}]}" > "${output_dir}/sums-current.txt"
+        echo "${images_sha[${j}]}" >> "${output_dir}/sums-new.txt"
+        if [[ "${url_current}" == "${url_new}" ]]; then
+            export_image "$url_new" "${output_dir}/image.img"
+        else
+            export_image "$url_new" "${output_dir}/image-new.img"
+            export_image "$url_current" "${output_dir}/image-current.img"
+            $delta_cmd "${output_dir}/image-current.img" "${output_dir}/image-new.img" "${output_dir}/image.img"
+            rm -f "${output_dir}/image-new.img" "${output_dir}/image-current.img"
+        fi
+    done
+}
+
+generate_metadata() {
+    local -r output="$1"
+    local sha
+    local -a shas
+
+    echo -ne "{" > "${output}"
+    echo -ne '"application_name":' >> "${output}"
+    printf '"%s",' "$application_name" >> "${output}"
+    echo -ne '"orchestrator":' >> "${output}"
+    printf '"%s",' "$orchestrator" >> "${output}"
+    echo -ne '"platform":' >> "${output}"
+    printf '"%s",' "$platform" >> "${output}"
+    echo -ne '"version":' >> "${output}"
+    printf '"%s",' "$version" >> "${output}"
+    echo -ne '"images":[' >> "${output}"
+    for sha in $(echo "${images_sha[@]}" | tr ' ' '\n' | sort | uniq); do
+        shas+=($sha)
+    done
+    for ((i = 0; i < ${#shas[@]}; i++)); do
+        printf '"%s"' "${shas[${i}]}" >> "${output}"
+        [[ $((i + 1)) -lt ${#shas[@]} ]] && echo -ne "," >> "${output}"
+    done
+    echo -ne ']' >> "${output}"
+    echo -ne "}" >> "${output}"
+}
+
+prepare_images "${temp_dir}/images"
+( cd "${temp_dir}" && tar czvf images.tar.gz images  )
+cp -a "${manifests_dir}" "${temp_dir}/manifests"
+( cd "${temp_dir}" && tar czvf "manifests.tar.gz" manifests  )
+generate_metadata "${temp_dir}/metadata.json"
+
+mender-artifact write module-image \
+    -T app \
+    $device_types \
```

Comment:
Nit: `quote`

Reply:
this was actually a copy-paste, and due to the way we pass the arguments it cannot be quoted, there is -t in there. I moved to a different approach, which I hope makes it better.